image/default_image_file.go

```diff
@@ -0,0 +1,156 @@
+		return fmt.Errorf("file %s already exists", imageTar)
+	}
+
+	if err := utils.MkFileFullPathDir(imageTar); err != nil {
+		return fmt.Errorf("failed to create %s, err: %v", imageTar, err)
+	}
+
+	return d.save(imageName, imageTar)
+}
+
+func (d DefaultImageFileService) Merge(image *v1.Image) error {
+	panic("implement me")
+}
+
+func (d DefaultImageFileService) save(imageName, imageTar string) error {
+	named, err := reference.ParseToNamed(imageName)
+	if err != nil {
+		return err
+	}
+
+	image, err := imageutils.GetImage(named.Raw())
+	if err != nil {
+		return err
+	}
+	file, err := os.Create(imageTar)
+	if err != nil {
+		return fmt.Errorf("failed to create %s, err: %v", imageTar, err)
+	}
+	defer file.Close()
+
+	var pathsToCompress []string
+	layerDirs, err := GetImageLayerDirs(image)
+	if err != nil {
+		return err
+	}
+	pathsToCompress = append(pathsToCompress, layerDirs...)
+	tempDir, err := utils.MkTmpdir()
+	if err != nil {
+		return fmt.Errorf("failed to create %s, err: %v", tempDir, err)
+	}
+	defer utils.CleanDir(tempDir)
+	imageMetadataTempFile := filepath.Join(tempDir, common.DefaultImageMetadataFileName)
+	repofile := filepath.Join(tempDir, common.DefaultMetadataName)
+	imageYaml := filepath.Join(common.DefaultImageMetaRootDir, image.Spec.ID+common.YamlSuffix)
+
+	ima, err := ioutil.ReadFile(imageYaml)
+	if err != nil {
+		return fmt.Errorf("failed to read %s, err: %v", imageYaml, err)
+	}
+	if err = ioutil.WriteFile(imageMetadataTempFile, ima, common.FileMode0644); err != nil {
+		return fmt.Errorf("failed to write temp file %s, err: %v ", imageMetadataTempFile, err)
+	}
+	repo, err := json.Marshal(&imageutils.ImageMetadata{ID: image.Spec.ID, Name: named.Raw()})
+	if err != nil {
+		return err
+	}
+	if err = ioutil.WriteFile(repofile, repo, common.FileMode0644); err != nil {
+		return fmt.Errorf("failed to write temp file %s, err: %v ", imageMetadataTempFile, err)
+	}
+
+	pathsToCompress = append(pathsToCompress, imageMetadataTempFile, repofile)
+	_, err = compress.Compress(file, pathsToCompress...)
+	return err
+}
+
+func (d DefaultImageFileService) load(imageSrc string) error {
+	src, err := os.Open(imageSrc)
+	if err != nil {
+		return fmt.Errorf("failed to open %s, err : %v", imageSrc, err)
+	}
+	defer src.Close()
+	if err = compress.Decompress(src, common.DefaultLayerDir); err != nil {
+		return err
+	}
+	repofile := filepath.Join(common.DefaultLayerDir, common.DefaultMetadataName)
+	defer os.Remove(repofile)
+	imageTempFile := filepath.Join(common.DefaultLayerDir, common.DefaultImageMetadataFileName)
+	defer os.Remove(imageTempFile)
+	var image v1.Image
+
+	if err = utils.UnmarshalYamlFile(imageTempFile, &image); err != nil {
+		return fmt.Errorf("failed to parsing %s.yaml, err: %v", imageTempFile, err)
+	}
+
+	layerStore, err := store.NewDefaultLayerStore()
+	if err != nil {
+		return fmt.Errorf("failed to get layerstore, err: %v", err)
+	}
+
+	for _, layer := range image.Spec.Layers {
+		if layer.Hash != "" {
```

Comment:
if layer.Hash == "" {continue}, do it in this way, which can help decrease code nesting